jl4-core/src/L4/Desugar.hs

```diff
@@ -0,0 +1,318 @@
+    MkObligation ann
+      <$> dsExpr party
+      <*> dsExpr action
+      <*> traverse dsExpr due
+      <*> traverse dsExpr hence
+      <*> traverse dsExpr lest
+
+dsLocalDecl :: LocalDecl Name -> Desugar (LocalDecl Name)
+dsLocalDecl = \ case
+  LocalDecide ann decide -> do
+    d <- dsDecide decide
+    pure $ LocalDecide ann d
+  LocalAssume ann assume -> do
+    pure $ LocalAssume ann assume
+
+dsBranch :: Branch Name -> Desugar (Branch Name)
+dsBranch = \ case
+  When ann pat e -> do
+    de <- dsExpr e
+    pure $ When ann pat de
+  Otherwise ann e -> do
+    de <- dsExpr e
+    pure $ Otherwise ann de
+
+-- | Rewrite the 'Anno' of the given arguments @'NonEmpty' ('Expr' 'Name)'@ to
+-- include the concrete syntax nodes of the 'Anno' in the @'Expr' 'Name'@.
+--
+-- Let's assume an 'Expr' for the code:
+--
+-- @
+--   1 PLUS 2
+-- @
+--
+-- We want to desugar this to use a prefix function:
+--
+-- @
+--   \_\_PLUS\_\_ 1 2
+-- @
+--
+-- To make exactprinting still faithful to the original sources, we need
+-- to be very careful that it is printed the same way. First, let's look at the
+-- 'Anno' for the @1 PLUS 2@ expression:
+-- (a @_@ marks an 'AnnoHole'):
+--
+-- @
+--   _ PLUS _
+-- @
+--
+-- The 'Anno' of a function looks like '_ [OF] _', so two holes, one for the name of the function
+-- and one for *all* arguments of the function. The first hole is easy to manage, just make
+-- sure the 'Name' of the function has an 'emptyAnno', then it is basically skipped over.
+-- For the second hole to accurately reproduce the same concrete syntax nodes, we need to be more
+-- careful.
+--
+-- We need to massage now the concrete syntax nodes of '_ PLUS _' into the arguments
+-- of the prefix function notation.
+-- For each 'AnnoHole' in the original 'Anno', we get all the 'AnnoCsn' that come
+-- afterwards and attach them to the 'Anno' of the argument.
+-- Note, for the first element, we need to do the same thing for leading 'AnnoCsn' elements.
+-- The finalised 'Anno' of each argument should then look like:
+--
+-- @
+--   [[_, " PLUS "], [_]]
+-- @
+--
+-- where the inner holes are then filled by the underlying @'Expr' 'Name'@.
+--
+-- This works also nicely for parenthesis, the expression @(1 PLUS 2)@
+-- is then translated to:
+--
+-- @
+--   [["(", _, " PLUS "], [_, ")"]]
+-- @
+rewriteBinOpAnno :: Expr Name -> Expr Name -> Expr Name -> Desugar [Expr Name]
+rewriteBinOpAnno expr e1 e2 = do
+  let
+
+  case csnSlices of
+    [beforeFirst, beforeSecond, after] ->
+      pure
+        [ updateAnnoOf e1 (surroundWithCsn beforeFirst beforeSecond)
+        , updateAnnoOf e2 (surroundWithCsn [] after)
+        ]
+    _slices -> internalError $ InternalAnnoRewritingError expr (HoleInfo 2 numberOfAnnoHoles)
+ where
+  annoPieces = (getAnno expr).payload
+
+  updateAnnoOf e f = e & annoOf %~ f
+
+  csnSlices = splitWhen (isJust . preview #_AnnoHole) annoPieces
+  numberOfAnnoHoles = length $ filter (isJust . preview #_AnnoHole) annoPieces
```

Comment:
this all seems very fragile. The docs make sense thought and I don't have a better idea at this moment. Maybe let's talk about this.  

I also don't think this function warrants an additional dependency...